_posts/2018-05-31-regular-types.md

```diff
@@ -0,0 +1,874 @@
+cost of pointers not being considered Regular because of semantic requirements
+and implementation-defined behavior on comparison. (If deletion is "as-if all
+other pointers to this object are scribbled over" that certainly violates the
+restriction on the value of the object changing out from under us.)
+
+All three of these options provide points in the type design space that are
+useful in conjunction with the existing definitions for Regular.
+
+-   **Thread-compatible** + Regular is what we really want for user-defined
+    types that mimic built-ins. This lets us reason about an instance in the
+    expected fashion and use it efficiently in conjunction with generic
+    algorithms. Types that have mutable data may have some overhead to support
+    this.
+-   **Dependent-preconditions** with knowledge that an instance + its dependent
+    data are safe to use. This is the common usage for `string_view` when we use
+    it as a non-owning parameter type: the underlying buffer will outlive the
+    function call and is immutable for the duration of the call. Given that
+    external knowledge of that underlying buffer, `string_view` behaves as if it
+    were Regular.
+-   **Single-threaded usage** - This is easy to misuse, but can be an important
+    area for optimization. Consider the discussions to provide a `shared_ptr`
+    analogue that does not synchronize its reference count - if we know
+    something about program structure, or can guarantee particular usage for an
+    instance, we can design a more efficient type in this fashion. Given that
+    knowledge, such a `shared_ptr` can still behave as if it were Regular.
+
+Given a type and one of the above options to explain why it is safe to use a
+given instance, we can see that our nice property for Regular types still holds.
+Given a `const T` (and some program knowledge), we can perform any operation
+(that conforms to that program knowledge) on that `const T` without invalidating
+it or any of its API preconditions. Regular+thread-compatible types allow us
+this invariant with no constraints on the program or that operation. Lesser
+invariants require more knowledge - in return we tend to get lower-overhead,
+which is a very C++ style of tradeoff.
+
+## Evaluating `string_view`
+
+If we know that the underlying buffer exists and is not being mutated,
+`string_view` behaves as if it were Regular: it is DefaultConstructible,
+CopyConstructible, Movable, Swappable, Assignable, EqualityComparable. If the
+underlying buffer is immutable for the life of our instance, the General Matter
+rules in P0898 don't kick in: the value won't magically change out from under
+us.
+
+```c++
+using T = string_view;
+void DoSomething(const T& t);
+
+const T a = SomeT();  // Assume SomeT() is providing a long-lived and stable buffer
+const T b = SomeT();
+if (a == b) {
+  DoSomething(a);  // Can't modify the buffer, it's assumed stable.
```

Comment:
Does this comment need to be clarified?

Can't modify the buffer directly or indirectly because we already assumed SomeT() is providing a stable underlying buffer.

Reply:
Changed to:

  DoSomething(a);     // Won't modify the buffer,
                                    // provided our assumption on SomeT() is correct.